pyscf/lib/dft/libxc_itrf2.c

```diff
@@ -0,0 +1,1147 @@
+        {
+                case XC_FAMILY_MGGA:
+#ifdef XC_FAMILY_HYB_MGGA
+                case XC_FAMILY_HYB_MGGA:
+#endif
+                        mgga = 1;
+                        break;
+                default:
+                        mgga = 0;
+        }
+
+        return mgga;
+}
+
+int LIBXC_needs_laplacian(const int nfunc, const xc_func_type *func)
+{
+        int i;
+        for (i = 0; i < nfunc; i++) {
+                if (func[i].info->flags & XC_FLAGS_NEEDS_LAPLACIAN) {
+                        return 1;
+                }
+        }
+        return 0;
+}
+
+int LIBXC_is_hybrid(const xc_func_type *func)
+{
+        int hyb;
+
+#if XC_MAJOR_VERSION <= 6
+        switch(func->info->family)
+        {
+#ifdef XC_FAMILY_HYB_LDA
+                case XC_FAMILY_HYB_LDA:
+#endif
+                case XC_FAMILY_HYB_GGA:
+                case XC_FAMILY_HYB_MGGA:
+                        hyb = 1;
+                        break;
+                default:
+                        hyb = 0;
+        }
+#else
+        hyb = (xc_hyb_type(func) == XC_HYB_HYBRID);
+#endif
+
+        return hyb;
+}
+
+double LIBXC_hybrid_coeff(const xc_func_type *func)
+{
+        double factor;
+
+#if XC_MAJOR_VERSION <= 6
+        switch(func->info->family)
+        {
+#ifdef XC_FAMILY_HYB_LDA
+                case XC_FAMILY_HYB_LDA:
+#endif
+                case XC_FAMILY_HYB_GGA:
+                case XC_FAMILY_HYB_MGGA:
+                        factor = xc_hyb_exx_coef(func);
+                        break;
+                default:
+                        factor = 0;
+        }
+
+#else
+        if(xc_hyb_type(func) == XC_HYB_HYBRID)
+                factor = xc_hyb_exx_coef(func);
+        else
+                factor = 0.0;
+#endif
+
+        return factor;
+}
+
+void LIBXC_nlc_coeff(const xc_func_type *func, double *nlc_pars) {
+
+        XC(nlc_coef)(func, &nlc_pars[0], &nlc_pars[1]);
+}
+
+void LIBXC_rsh_coeff(const xc_func_type *func, double *rsh_pars) {
+
+        rsh_pars[0] = 0.0;
+        rsh_pars[1] = 0.0;
+        rsh_pars[2] = 0.0;
+
+#if XC_MAJOR_VERSION <= 6
+        XC(hyb_cam_coef)(func, &rsh_pars[0], &rsh_pars[1], &rsh_pars[2]);
+#else
+        switch(xc_hyb_type(func)) {
+        case(XC_HYB_HYBRID):
+        case(XC_HYB_CAM):
+                XC(hyb_cam_coef)(func, &rsh_pars[0], &rsh_pars[1], &rsh_pars[2]);
+        }
+#endif
+}
+
+int LIBXC_is_cam_rsh(const xc_func_type *func) {
+#if XC_MAJOR_VERSION <= 6
+        return func->info->flags & XC_FLAGS_HYB_CAM;
+#else
+        return (xc_hyb_type(func) == XC_HYB_CAM);
+#endif
+}
+
+/*
+ * XC_FAMILY_LDA           1
+ * XC_FAMILY_GGA           2
+ * XC_FAMILY_MGGA          4
+ * XC_FAMILY_LCA           8
+ * XC_FAMILY_OEP          16
+ * XC_FAMILY_HYB_GGA      32
+ * XC_FAMILY_HYB_MGGA     64
+ * XC_FAMILY_HYB_LDA     128
+ */
+int LIBXC_xc_type(const int nfunc, const xc_func_type *func)
+{
+        int i;
+        int type = -1;
+        for (i = 0; i < nfunc; i++) {
```

Comment:
This is because in the new implementation, functionals are processed in units of each `xc_code`, rather than in each integer ID. It would be clearer (and might be slightly more efficient) to process all the functional components of each `xc_code` at once in the C layer whenever possible.